qiskit/circuit/singleton.py

```diff
@@ -0,0 +1,250 @@
+    __slots__ = ()
+
+    def __new__(mcs, name, bases, namespace, *, overrides=None, **kwargs):
+        cls = super().__new__(mcs, name, bases, namespace, **kwargs)
+        if overrides is not None:
+            # The `__init_subclass__` logic is shared between `SingletonInstruction` and
+            # `SingletonGate`, but we can't do `__init_subclass__ = _impl_init_subclass(...)`
+            # inside the class definition bodies because it couldn't make `super()` calls.
+            cls.__init_subclass__ = _impl_init_subclass(cls, overrides)
+        return cls
+
+    def __call__(cls, *args, _force_mutable=False, **kwargs):
+        if not _force_mutable and not args and not kwargs:
+            # This class attribute is created by the singleton-creation base classes'
+            # `__init_subclass__` methods; see `_impl_init_subclass`.
+            return cls._singleton_instance
+        return super().__call__(*args, **kwargs)
+
+
+class _SingletonInstructionOverrides(Instruction):
+    """Overrides for all the mutable methods and properties of `Instruction` to make it
+    immutable."""
+
+    __slots__ = ()
+
+    def c_if(self, classical, val):
+        return self.to_mutable().c_if(classical, val)
+
+    @property
+    def base_class(self):
+        # `type(self)` will actually be the dynamic `_SingletonXGate` (e.g.) created by
+        # `SingletonGate.__init_subclass__` during the instantiation of `XGate`, since this class
+        # is never the concrete type of a class.
+        return type(self)._base_class
+
+    @property
+    def mutable(self):
+        return False
+
+    def to_mutable(self):
+        return self.base_class(_force_mutable=True)
+
+    def copy(self, name=None):
+        if name is None:
+            return self
+        out = self.to_mutable()
+        out.name = name
+        return out
+
+    def __setattr__(self, key, value):
+        raise NotImplementedError(
+            f"This '{self.base_class.__name__}' object is immutable."
+            " You can get a mutable version by calling 'to_mutable()'."
+        )
+
+    def __copy__(self):
+        return self
+
+    def __deepcopy__(self, _memo=None):
+        return self
+
+    def __reduce__(self):
+        return (operator.attrgetter("_singleton_instance"), (self.base_class,))
+
+
+class SingletonInstruction(
+    Instruction, metaclass=_SingletonMeta, overrides=_SingletonInstructionOverrides
+):
+    """A base class to use for :class:`~.circuit.Instruction` objects that by default are singleton
+    instances.
+
+    This class should be used for instruction classes that have fixed definitions and do not contain
+    any unique state. The canonical example of something like this is :class:`.Measure` which has an
+    immutable definition and any instance of :class:`.Measure` is the same. Using singleton
+    instructions as a base class for these types of gate classes provides a large advantage in the
+    memory footprint of multiple instructions.
```

Comment:
Just something I'm thinking about based on our offline discussion about this PR we were talking about the case of something like:

```
class ESPMeasure(Measure)
```

For the purposes of having specialized measurement operations in the target. How does this pattern work with regards to singletons? It's honestly something I hadn't considered with `SingletonGate` originally and it's not a blocker, but I think knowing what our expectations are here will be good.

Reply:
Quick answer:
```python
from qiskit.circuit.singleton import SingletonInstruction

class Measure(SingletonInstruction):
    def __init__(self, label=None, *, duration=None, unit="dt"):
        return super().__init__("measure", 1, 1, (), duration=duration, unit=unit, label=label)

class ESPMeasure(Measure):
    pass

measure = Measure()
esp = ESPMeasure()

assert measure is Measure()
assert esp is ESPMeasure()
assert measure is not esp

assert measure.base_class is Measure
assert esp.base_class is ESPMeasure
```
passes.

Long answer:

Since `ESPMeasure` inherits `Measure`, it also inherits `SingletonInstruction`, which means the `__init_subclass__` runs when `ESPMeasure` is created, and its metaclass is `_SingletonMeta` so it'll use that to create the object.

Stepping through:
1. The `ESPMeasure` name is created by
   ```python
   ESPMeasure = _SingletonMeta.__new__(_SingletonMeta, "ESPMeasure", (Measure,), <executed class body namespace>, **{})
   ```
   where the last argument is empty because we didn't specify any kwargs.  I'm ignoring anything that might be defined in the class body - it just gets executed into the penultimate argument in a class scope (standard stuff).

  `_SingletonMeta.__new__` doesn't do anything if the `overrides` kwarg isn't given, so this just passes everything to `abc.ABCMeta.__new__`, which in turn passes it to `type.__new__`, so we get a class created completely regularly.
2. Now the resulting object is a subclass of `SingletonInstruction`, which has an `__init_subclass__`, so now we call that.  Again, there were no kwargs in the class-definition statement, so notably, `create_singleton=True` by default.  The call ends up looking like this:
  ```python
  ESPMeasure = _SingletonMeta.__new__(...)  # from before
  ESPMeasure = ESPMeasure.__init_subclass__()
  ```
  so in the defined `__init_subclass__`, `cls=ESPMeasure`, `create_singleton=True` and since it's the same closure from `SingletonInstruction`, `base=SingletonInstruction`.  Now this dynamically creates the type object `_SingletonESPMeasure` with the bases `(_SingletonInstructionOverrides, ESPMeasure)`, and `_base_class=ESPMeasure`, so it'll return the right thing.

Reply:
I will write a test of this, though - I completely spaced on that.

Reply:
Added documentation on this in 145decf53e, and tests in 1274fe79ce.